DXMainClient/Domain/Multiplayer/CnCNet/CnCNetTunnel.cs

```diff
@@ -85,6 +85,59 @@ public static CnCNetTunnel Parse(string str)
         public double Distance { get; private set; }
         public int PingInMs { get; set; } = -1;
 
+        /// <summary>
+        /// Rating Factor:
+        ///   The latency in milleseconds that ping times are rounded off to.
+        /// </summary>
+        private const int LATENCY_PRECISION =  70;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The importance of latency compared to client count.
+        /// </summary>
+        private const int LATENCY_WEIGHT = 100000;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The value of client count compared to latency
+        /// </summary>
+        private const int CLIENTS_WEIGHT = 1;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   Base Rating for untrusted servers and unpingable servers.
+        /// </summary>
+        private const int UNTRUSTED_RATING = int.MaxValue - 100000;
+
+        /// <summary>
+        /// Rating Factor:
+        ///   The amount of free client slots needed for the tunnel to be eligible for autoselection
+        /// </summary>
+        private const int CLIENTS_HEADROOM = 24;
+
+        public int Rating
+        {
+            get
+            {
+                if (Clients + CLIENTS_HEADROOM >= MaxClients)
+                    return int.MaxValue;
+
+                if (Official || Recommended)
+                {
+                    if (PingInMs <= -1)
+                        return UNTRUSTED_RATING + Clients;
+
+                    int latency = 1 + PingInMs / LATENCY_PRECISION;
+
+                    return (latency * LATENCY_WEIGHT) + (Clients * CLIENTS_WEIGHT);
+                }
+                else
+                {
+                    return UNTRUSTED_RATING + Clients;
```

Comment:
With this code it looks like the clients would almost never get an unofficial and non-recommended server selected. This is not the behavior I see as good, CnCNet is meant to be a community service, and IMO recommendation should boost the server, not basically "whitelist" non-recommended servers.

Reply:
This is the best case, the recommended servers have all been tested and proven to be good. The goal is for the users to never want or need to select a tunnel server themselves. When they do manually select a tunnel server it should be their own server that they want to use for personal reasons.  This provides the added benefit that we can do some hacks on the master-list side to balance tunnel servers during the high load times.

Reply:
I still object against verified and official tunnels being effectively whitelisted against regular ones, this is not future-proof at all. I mean the example is already at hand. I have a feeling no one actually bothered to test Berryville, yet it has very low ping for Europe and I never had issues using it. Imagine CnCNet staff going very busy all at once, then some of the verified tunnel will also go off and the players will be automatically balanced to 2-3 verified servers in USA, for example, even if they are from EU and there is regular, non-verified EU servers. This is absolutely not a desirable solution.

Reply:
The reason we should never use untrusted tunnel servers is because they fail more often. When a tunnel server fails and it breaks the game for a group of players, the players don't blame the tunnel server owner they blame CnCNet, we can't stake our reputation on untrusted servers.

> Imagine CnCNet staff going very busy all at once, then some of the verified tunnel will also go off and the players will be automatically balanced to 2-3 verified servers in USA, for example, even if they are from EU and there is regular, non-verified EU servers. This is absolutely not a desirable outcome IMO.

That's still more desirable than a broken tunnel server with low ping destroying everyone's game.

Reply:
My view of future CnCNet is as of a community service, not an organization, so I don't think this outcome is more desirable. I don't say that the system should be ditched. Per my idea verified servers would be "boosted" and chosen most of the time, but in case there's no local verified servers with low enough ping - regular one would be chosen. This would never happen in the current implementation.

Again, you say that untrusted servers are bad because all good servers are made verified. Is that so? Not really. Someone has to test them manually, and as I said I don't see anyone testing Berryville server or some other EU servers which are OK despite the lack of EU tunnels. I always get Canada East chosen and I live in Europe.